### PR TITLE
prov/efa: Enable `FI_HMEM_CUDA` usage when p2p is unavailable

### PR DESCRIPTION
**Pull request by darrylabbate**
~~NOTE: Marking as "Do not merge" until #8357 is merged into `main` and this PR's branch can be rebased.~~ #8357 merged

This implements various minor changes to enable the `FI_HMEM_CUDA` interface even when GPUDirect RDMA is unavailable.

### CODE REVIEW

File: prov/efa/src/efa_hmem.c
Comment (wzamazon) [inline prov/efa/src/efa_hmem.c:140]: This is not quite right. We still need to check `cuda_get_xfer_setting()`. Only when `cuda_get_xfer_setting()` is `CUDA_XFER_ENABLED` we do not require p2p by implemenation.

File: prov/efa/src/efa_user_info.c
Comment (wzamazon) [inline prov/efa/src/efa_user_info.c:347]: This depend on the value of `cuda_get_xfer_setting()`.

If `cuda_get_xfer_setting()` is CUDA_XFER_ENABLED. we support FI_HMEM.
Otherwise, we support FI_HMEM with p2p + read.

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/src/efa_mr.c
Comment (wzamazon) [inline prov/efa/src/efa_mr.c:63]: I would just use macro .e. `#define`

File: prov/efa/src/rxr/rxr_pkt_type_base.c
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_base.c:409]: I need to think about this change more.

File: prov/efa/src/efa_hmem.c
Comment (darrylabbate) [reply prov/efa/src/efa_hmem.c:140]: Based on your changes in #8357, it would have to be `FI_HMEM_CUDA_ENABLE_XFER=1` or `FI_HMEM_CUDA_USE_GDRCOPY=1`. Otherwise there'd be no path to use GDRCopy.

File: prov/efa/src/efa_mr.c
Comment (darrylabbate) [reply prov/efa/src/efa_mr.c:63]: ack

File: prov/efa/src/efa_user_info.c
Comment (darrylabbate) [reply prov/efa/src/efa_user_info.c:347]: Same issue here, if we restrict `FI_HMEM_CUDA` to just `FI_HMEM_CUDA_ENABLE_XFER=1`, there's no GDRCopy path since #8357 restricts `cuda_gdrcopy_dev_register` to the following conditions `efa_mr_hmem_setup()`:

```c
static inline
bool efa_mr_use_gdrcopy(struct efa_mr *efa_mr)
{
	return efa_mr->peer.iface == FI_HMEM_CUDA && cuda_get_xfer_setting()!=CUDA_XFER_ENABLED && cuda_is_gdrcopy_enabled();
}
```

Comment (wzamazon) [reply prov/efa/src/efa_user_info.c:347]: Yes. that is intented. we do not want to use gdrcopy to support HMEM on non-p2p systems. (performance is an issue, always data consistency is another).

File: prov/efa/src/efa_hmem.c
Comment (wzamazon) [reply prov/efa/src/efa_hmem.c:140]: We do not want to use gdrcopy to support non-p2p.

File: prov/efa/src/efa_user_info.c
Comment (darrylabbate) [reply prov/efa/src/efa_user_info.c:347]: I see. This is news to me, but not a big deal. I can easily add the restriction on `p2p_required_by_impl` to `cuda_get_xfer_setting() == CUDA_XFER_ENABLED`.

File: prov/efa/src/rxr/rxr_pkt_type_base.c
Comment (darrylabbate) [reply prov/efa/src/rxr/rxr_pkt_type_base.c:409]: Actually if we're abandoning GDRCopy no non-p2p, this commit can be discarded altogether. Unless we want to keep it in to support later.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_base.c:409]: In that case, let's remove this commit from this PR first. I will exam the code, and think about it more.

File: prov/efa/src/efa_user_info.c
Comment (darrylabbate) [reply prov/efa/src/efa_user_info.c:347]: So which of the original conditions should be kept in? `!efa_device_support_rdma_read()`? E.g.

```c
if (!efa_device_support_rdma_read() && cuda_get_xfer_setting() != CUDA_XFER_ENABLED) {
    EFA_WARN(...);
    return -FI_ENODATA;
}
```

File: prov/efa/src/efa_user_info.c
Comment (darrylabbate) [inline prov/efa/src/efa_user_info.c:330]: How do we feel about this warning message? Would it be a good idea to add something about enable `FI_HMEM_CUDA_ENABLE_XFER` if the user wishes to use non-p2p?

Comment (wzamazon) [reply prov/efa/src/efa_user_info.c:330]: Yes. I think we need update  the warning message to add FI_HMEM_CUDA_ENABLE_XFER

Comment (wzamazon) [reply prov/efa/src/efa_user_info.c:330]: You dropped rxr_env.use_device_rdma.

Comment (darrylabbate) [reply prov/efa/src/efa_user_info.c]: ack, see latest push

File: prov/efa/src/efa_mr.c
Comment (wzamazon) [inline prov/efa/src/efa_mr.c:824]: add `efa_mr->ibv_mr =  NULL`  to be safe

File: prov/efa/src/efa_user_info.c
Comment (wzamazon) [inline prov/efa/src/efa_user_info.c:334]: indentation seems off.

**Review by wzamazon — COMMENTED** [review]
Overall looks good to me. Except some minor comments. Thank you!

File: prov/efa/src/efa_mr.c
Comment (darrylabbate) [reply prov/efa/src/efa_mr.c:824]: `efa_mr->ibv_mr` is initialized to `NULL` on line 792. In what scenario could `efa_mr->ibv_mr` be anything other than `NULL` when it hits this conditional?

Comment (wzamazon) [reply prov/efa/src/efa_mr.c:824]: Was not line 792 removed?

File: prov/efa/src/efa_user_info.c
Comment (darrylabbate) [reply prov/efa/src/efa_user_info.c]: ack, reduced the extra indentations in this file

**Review by wzamazon — APPROVED** [review]

**Comment by darrylabbate** [comment]
- Rebased branch after #8357 merged into `main`
- Cherry-picked 3e29de1d16313bb05d712d0401c131819d6eef18 from @wzamazon and reordered commits s.t. it precedes the commit where the p2p/RDMA requirements are loosened - this keeps the commit history stable
- Tested by running `osu_bw -d cuda D D` on 2 g4dn.12xlarge nodes with `FI_HMEM_CUDA_ENABLE_XFER=1` to confirm everything works as expected

**Comment by wzamazon** [comment]
Thank you!

waiting for CI to complete. Then will merge

**Pull request merged**
